### PR TITLE
Skip ALL granularity compaction 

### PR DESCRIPTION
**Pull request by AmatyaAvadhanula**
Fixes https://github.com/apache/druid/issues/13208

### Description

Skip datasources with ALL granularity segments for auto-compaction

<hr>

##### Key changed/added classes in this PR
 * `NewestSegmentFirstIterator`

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

File: server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java [hunk 1]
```diff
@@ -129,6 +131,12 @@ public class NewestSegmentFirstIterator implements CompactionSegmentIterator
             // For example, if the original is interval of 2020-01-28/2020-02-03 with WEEK granularity
             // and the configuredSegmentGranularity is MONTH, the segment will be split to two segments
             // of 2020-01/2020-02 and 2020-02/2020-03.
+            if (Intervals.ETERNITY.equals(segment.getInterval())
+                && !Granularities.ALL.equals(configuredSegmentGranularity)) {
+              log.warn("Cannot compact segments from ALL to finer granularity for datasource[%s]", dataSource);
```
Comment (gianm) [inline server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java:136]: Please include a link to this PR, so future devs can understand why we can't do this. One day, hopefully someone will fix it.

File: server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java [hunk 2]
```diff
@@ -129,6 +131,12 @@ public class NewestSegmentFirstIterator implements CompactionSegmentIterator
...
+              allToFinerGranularityCompaction = true;
+              continue;
```
Comment (gianm) [inline server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java:138]: Early return directly from here?

Comment (AmatyaAvadhanula) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java]: Done

File: server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java (hunk 1 above)
Comment (AmatyaAvadhanula) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java:135]: https://github.com/apache/druid/issues/13208#issuecomment-1302886876

**Comment by AmatyaAvadhanula** [comment]
@gianm, thank you for the review.

I've addressed your feedback, and will also try to enable compaction from ALL granularity in subsequent PRs

**Comment by abhishekagarwal87** [comment]
@AmatyaAvadhanula - Please add this gotcha to the docs as well.

File: server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java (hunk 1 above)
Comment (abhishekagarwal87) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java:135]: What frequency will this be logged at?

Comment (AmatyaAvadhanula) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java:135]: Once every `druid.coordinator.period.indexingPeriod` or coordinator custom compaction period, whichever is more frequent

**Comment by AmatyaAvadhanula** [comment]
@abhishekagarwal87 thank you for the feedback. I've added a note in the auto-compaction docs

**Review by abhishekagarwal87 — APPROVED** [review]

File: docs/data-management/automatic-compaction.md [hunk 3]
```diff
@@ -166,6 +166,8 @@ The following auto-compaction configuration compacts existing `HOUR` segments in
 }
 ```
 
+Please note that auto-compaction currently skips ALL granularity intervals when the target segment granularity is finer.
```
Comment (kfaraz) [inline docs/data-management/automatic-compaction.md:169]: Maybe put this as a quoted section:
e.g.
```suggestion
> Auto-compaction skips segments with granularity ALL if the target segment granularity is not ALL.
```

File: server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java (hunk 1 above)
Comment (AmatyaAvadhanula) [reply server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java:136]: Thanks @kfaraz. The custom compaction duty's frequecy is chosen

File: server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java [hunk 4]
```diff
@@ -129,6 +130,12 @@ public class NewestSegmentFirstIterator implements CompactionSegmentIterator
             // For example, if the original is interval of 2020-01-28/2020-02-03 with WEEK granularity
             // and the configuredSegmentGranularity is MONTH, the segment will be split to two segments
             // of 2020-01/2020-02 and 2020-02/2020-03.
+            if (Intervals.ETERNITY.equals(segment.getInterval())
+                && !Granularities.ALL.equals(configuredSegmentGranularity)) {
+              // This is to prevent the coordinator from crashing as raised in https://github.com/apache/druid/issues/13208
+              log.warn("Cannot compact segments from ALL to finer granularity for datasource[%s]", dataSource);
```
Comment (cryptoe) [inline server/src/main/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstIterator.java:136]:
```suggestion
              log.warn("Cannot compact segments from ALL to finer granularity[%s] for datasource[%s]", configuredSegmentGranularity, dataSource);
```

**Review by cryptoe — COMMENTED** [review]
Minor comments

File: server/src/test/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstPolicyTest.java [hunk 5]
```diff
@@ -1584,6 +1585,38 @@ public void testIteratorDoesNotReturnSegmentWithChangingAppendableIndexSpec()
     Assert.assertFalse(iterator.hasNext());
   }
 
+  @Test
+  public void testAllToFinerGranularity()
```
[the remark this answers is not in the document]
Comment (AmatyaAvadhanula) [reply server/src/test/java/org/apache/druid/server/coordinator/duty/NewestSegmentFirstPolicyTest.java]: @cryptoe thanks for the feedback. ALL granularity segments in general don't seem compatible with autocompaction.
The patch now skip such datasources and relevant tests have been added

**Comment by abhishekagarwal87** [comment]
@AmatyaAvadhanula what exactly happens in ALL -> ALL compaction scenario?

**Comment by AmatyaAvadhanula** [comment]
@abhishekagarwal87 the compaction tasks fail: 
`org.apache.druid.java.util.common.IAE: Granularity is not supported. [P292277024Y7M2W2DT7H12M55.807S]`

**Comment by AmatyaAvadhanula** [comment]
Merging since test failure is due to known flaky test: https://github.com/apache/druid/issues/12795

**Pull request merged**